test/jdk/tools/jlink/basic/AllModulePath.java

```diff
@@ -84,60 +102,141 @@ public void setup() throws Throwable {
+    public void noModulePath() throws IOException {
+        if (isExplodedJDKImage()) {
+            return;
+        }
+        Path targetPath = HELPER.createNewImageDir("all-mod-path-no-mod-path");
+        List<String> allArgs = List.of("--add-modules", "ALL-MODULE-PATH",
+                                       "--output", targetPath.toString());
+        JlinkOutput allOut = createImage(targetPath, allArgs, false /* expect failure */);
+        String expected = "Error: --module-path option must be specified with --add-modules ALL-MODULE-PATH";
+        assertEquals(allOut.stdout.trim(), expected);
+    }
+
+    /*
+     * --module-path not-exist and --add-modules ALL-MODULE-PATH is an error.
+     */
+    @Test
+    public void modulePathEmpty() throws IOException {
+        if (isExplodedJDKImage()) {
+            return;
+        }
+        Path targetPath = HELPER.createNewImageDir("all-mod-path-not-existing");
+        String strNotExists = "not-exist";
+        Path notExists = Path.of(strNotExists);
+        if (Files.exists(notExists)) {
+            throw new RuntimeException("Test setup error, path must not exist!");
+        }
+        List<String> allArgs = List.of("--add-modules", "ALL-MODULE-PATH",
+                                       "--module-path", notExists.toString(),
+                                       "--output", targetPath.toString());
+
+        JlinkOutput allOut = createImage(targetPath, allArgs, false /* expect failure */);
+        String actual = allOut.stdout.trim();
+        assertTrue(actual.startsWith("Error: No module found in module path"));
+        assertTrue(actual.contains(strNotExists));
+    }
+
+    /*
+     * --add-modules ALL-MODULE-PATH with an existing module path and module
+     * limits applied. This case test a module limit on a dependency, jdk.jfr,
+     * and *doesn't* list the module explicitly in --add-modules. Therefore,
+     * expects for the module - on the module path - to be not be present.
+     */
+    @Test
+    public void modulePathWithLimitMods() throws Exception {
+        if (isExplodedJDKImage()) {
+            return;
+        }
+        Path targetPath = HELPER.createNewImageDir("all-mods-limit-mods");
+        String moduleName = "com.baz.runtime";
+        Result result = HELPER.generateDefaultJModule(moduleName, "jdk.jfr");
+        Path customModulePath = result.getFile().getParent();
+        List<String> allArgs = List.of("--add-modules", "ALL-MODULE-PATH",
+                                       "--limit-modules", "jdk.jfr", // A dependency of com.baz.runtime
+                                       "--module-path", customModulePath.toString(),
+                                       "--output", targetPath.toString());
+        JlinkOutput allOut = createImage(targetPath, allArgs, true /* success */);
+        assertTrue(allOut.stdout.isEmpty());
+        assertTrue(allOut.stderr.isEmpty());
+        List<String> expected = List.of("java.base", "jdk.jfr");
```

Comment:
```
jlink --add-modules ALL-MODULE-PATH --limit-modules jdk.jfr --module-path jmods --output all-mods-limit-mods.image 
```

This PR changes to find the root modules from the paths specified in `--module-path` only.   So I would expect that `jdk.jfr` should not be included in the resulting image because it's not on `--module-path`.    This seems a bug as it finds all modules using the module finder that includes `jdk.jfr`.   The module finder should be:

```
                ModuleFinder mf = newLimitedFinder(finder,
                                                   options.limitMods.isEmpty() ? initialRoots : Set.of(),
                                                   Set.of());
```

Reply:
> This PR changes to find the root modules from the paths specified in --module-path only.

Yes, but in this case modules on the module path contain modules that require `jdk.jfr`:

```
$ jmod describe ./build/linux-x86_64-server-release/test-support/jtreg_test_jdk_tools_jlink_basic_AllModulePath_java/scratch/0/jmods/com.baz.runtime.jmod 
com.baz.runtime
exports com.baz.runtime
exports com.baz.runtime.com.foo.bar
requires java.base mandated
requires jdk.jfr
main-class com.baz.runtime.Main
```

So I think it's OK to limit to a smaller set of modules even though it seems a strange use-case.

A different question would be what should happen if the module specified in `--limit-modules` in **not** in the dependency graph of ALL-MODULE-PATH modules.

Reply:
> A different question would be what should happen if the module specified in `--limit-modules` is **not** in the dependency graph of ALL-MODULE-PATH modules.

This seems a good candidate for an error as it's unexpected, IMO. I'll add a separate test case which exercises this and add code to prevent this:

```
$ rm -rf ./build/limit-mods-test && ./build/linux-x86_64-server-release/images/jdk/bin/jlink --add-modules ALL-MODULE-PATH --module-path ./build/jmods --limit-modules jdk.net --verbose --output ./build/limit-mods-test
Error: jdk.net not in module dependency graph, but specified with --limit-modules
```

The same without limit modules:
```
$ rm -rf ./build/limit-mods-test && ./build/linux-x86_64-server-release/images/jdk/bin/jlink --add-modules ALL-MODULE-PATH --module-path ./build/jmods --verbose --output ./build/limit-mods-test
Linking based on the current run-time image
com.bar.testmod file:///disk/openjdk/upstream-sources/git/jdk-jdk/./build/jmods/com.bar.testmod.jmod
java.base jrt:/java.base (run-time image)
java.compiler jrt:/java.compiler (run-time image)
jdk.compiler jrt:/jdk.compiler (run-time image)
jdk.internal.opt jrt:/jdk.internal.opt (run-time image)
jdk.jdeps jrt:/jdk.jdeps (run-time image)
jdk.zipfs jrt:/jdk.zipfs (run-time image)

Providers:
  jdk.compiler provides com.sun.tools.javac.platform.PlatformProvider used by jdk.compiler,jdk.jdeps
  java.base provides java.nio.file.spi.FileSystemProvider used by java.base
  jdk.zipfs provides java.nio.file.spi.FileSystemProvider used by java.base
  jdk.compiler provides java.util.spi.ToolProvider used by java.base
  jdk.jdeps provides java.util.spi.ToolProvider used by java.base
  jdk.compiler provides javax.tools.JavaCompiler used by java.compiler
  jdk.compiler provides javax.tools.Tool not used by any observable module
```

A similar - pre-existing - problem exists where `--limit-modules` would be outside the dependency graph of modules specified with `--add-modules`. But this would be for a different bug so as to prevent scope-creep.

Reply:
> But this would be for a different bug so as to prevent scope-creep.

Filed https://bugs.openjdk.org/browse/JDK-8346299 for this.

Reply:
```
jlink --add-modules ALL-MODULE-PATH --limit-modules jdk.jfr --module-path jmods --output all-mods-limit-mods.image
```

The set of observable modules only has `java.base` and `jdk.jfr` because of `--limit-modules`.   Hence `com.baz.runtime` from `jmods` is not observable.    `ALL-MODULE-PATH` adds all observable modules found on the relevant module paths are added to the root set.   This matches the run-time behavior.

```
$ java --list-modules -p com.baz.runtime.jar --add-modules ALL-MODULE-PATH --limit-modules jdk.jfr
java.base@25-internal
jdk.jfr@25-internal
```

Reply:
Limiting the observable modules is sometimes useful for testing/debugging and to reduce the number of modules resolved when the main module is the unnamed module.

@AlanBateman  Is `--limit-modules` needed at link time?

Reply:
IMO it's also useful to limit included modules when `--bind-services` is being used.

Reply:
@mlchung It's not clear if you think there is an action expected from myself. If there is, please clarify what. Thanks!